### DISCUSSION TITLE
socket.io 에서 disconnect랑 disconnecting 부분 취합하기로 한건가요?

### DISCUSSION THREAD

**Discussion by tastekim** [tastekim/WeAllLie-BE#156] [Q&A]
#154 #153  PR 부분에서 각자 `disconnect`랑 `disconnecting`에 기능별로 socket 연결이 끊겼을 때 로직이 짜여있는데 각자 따로 가기로 한 건가요?
@kimhsno1 @Minhyeong-S

**Comment by ghost** [comment]
처음에는 그랬는데 어제 밤 이후로 아직 결정된건 없습니다!! 
근데 지금의 생각은 어제 연석님이 말씀하신 게 맞는 것 같아요!
disconnect나 disconnecting을 사용하는게 어짜피 room이든 chat이든 비정상적으로 room 에서 연결 해제됐을 때에 대한 대응인데,
같은 결의 이벤트니까 한 곳에서 처리하는게 좋은 것 같습니다. 
disconnecting 을 꼭 써야만 하는 이유가 있는 것도 아니니까..
(기존에 chat에서 disconnecting을 사용했던 이유는 room 확인이었는데요, 처음에 프론트로부터 roomNum 을 받았을 때 `socket.roomNum = roomNum` 이렇게 저장하는 방식으로 변경되어서 연결해제 상태에서도 roomNum을 확인할 수 있게 되어서 지금은 필요 없는 듯)

```
if (socket.rooms.size > 1) {
    const nickname = socket.nickname;
    const.roomNum = socket.roomNum
    console.log(`${nickname} 방 퇴장`);
    const msg = `${nickname} 님이 퇴장하셨습니다.`;
    // 방법1
    socket.to(`/gameRoom${roomNum}`).emit('receiveRoomMsg', { notice: msg });
    // 방법2
    // chat.to(`/gameRoom${roomNum}`).emit('receiveRoomMsg', { notice: msg });
    // 방법3
    // chat.in(`/gameRoom${roomNum}`).emit('receiveRoomMsg', { notice: msg });
}
```
채팅 쪽에서는 이 부분만 추가하면 되기는 한데 근데 지금 프론트에서 receiveRoomMsg를 못 받고 있는 이슈가 있어서,
서버에서도 방법을 여러가지로 시도해보고 있었거든요.. (어제는 근데 셋 다 실패 😭 :)
아직 확정이 아니라서 나중에 제대로 작동할 때 채팅 쪽 코드 추가해달라고 요청하겠습니다!

(추가)
그리고 저건 disconnecting 기준으로 작성했던 코드라서 `socket.rooms.size` 로 조건을 확인했었는데,
disconnect에 통합한다면 disconnect 에서 사용하는 **_유저가 방에 있었던 상태를 특정하는 로직_** 내에서 적용하면 될 것 같습니다.

**Reply by tastekim** [reply]
지금 형석님 PR에 `disconnect`부분에 첫 줄 부분인데 socket.roomNum으로 socket이 끊기는 그 시점에서 socket이 포함되어 있던 방의 id값을 socket안에 넣어놨습니다. 혹시 이걸로도 민형님 로직에 대체할 수 있을까요?

```javascript
    socket.on('disconnect', async () => {
        if (socket.roomNum === undefined || socket.roomNum === null) {
            userCnt--;
            lobby.emit('userCount', userCnt);
```

**Reply by ghost** [reply]
넵!! else 부분에 
```
socket.to(`/gameRoom${roomNum}`).emit('receiveRoomMsg', { notice: msg });
```
 가 들어가는 걸로 같이 확인!!
